File: app/streamlit_reply_agent/reply_gate.test.py
"""Tests for reply_gate recovery threshold."""

from __future__ import annotations

import unittest

from reply_gate import (
    RECOVERY_CONFIDENCE_THRESHOLD,
    apply_reply_gate,
    is_recovery_interest_tag,
)


class ReplyGateTests(unittest.TestCase):
    def test_recovery_tags(self) -> None:
        self.assertTrue(is_recovery_interest_tag(None))
        self.assertTrue(is_recovery_interest_tag(0))
        self.assertTrue(is_recovery_interest_tag(-1))
        self.assertFalse(is_recovery_interest_tag(1))
        self.assertFalse(is_recovery_interest_tag(-4))

    def test_interested_bypasses_threshold(self) -> None:
        result = apply_reply_gate(
            1,
            {
                "should_reply": True,
                "reply_text": "Bonjour",
                "reason": "ok",
                "recovery_confidence": 10,
            },
        )
        self.assertTrue(result["allow_reply"])

    def test_recovery_below_threshold(self) -> None:
        result = apply_reply_gate(
            -1,
            {
                "should_reply": True,
                "reply_text": "Bonjour",
                "reason": "low",
                "recovery_confidence": RECOVERY_CONFIDENCE_THRESHOLD - 1,
            },
        )
        self.assertFalse(result["allow_reply"])
        self.assertEqual(result["ai_status"], "skipped_recovery")

    def test_recovery_at_threshold(self) -> None:
        result = apply_reply_gate(
            -1,
            {
                "should_reply": True,
                "reply_text": "Bonjour",
                "reason": "ok",
                "recovery_confidence": RECOVERY_CONFIDENCE_THRESHOLD,
            },
        )
        self.assertTrue(result["allow_reply"])


if __name__ == "__main__":
    unittest.main()

File: lib/ai-reply-agent/reply-gate.test.ts
import { describe, expect, it } from "vitest";

import {
  applyReplyGate,
  isRecoveryInterestTag,
  NOT_INTERESTED_STATUS,
  RECOVERY_CONFIDENCE_THRESHOLD,
} from "./reply-gate";

describe("isRecoveryInterestTag", () => {
  it("treats Lead and Not interested as recovery tags", () => {
    expect(isRecoveryInterestTag(null)).toBe(true);
    expect(isRecoveryInterestTag(NOT_INTERESTED_STATUS)).toBe(true);
    expect(isRecoveryInterestTag(1)).toBe(false);
    expect(isRecoveryInterestTag(-4)).toBe(false);
  });
});

describe("applyReplyGate", () => {
  it("allows Interested without confidence gate", () => {
    const result = applyReplyGate(1, {
      should_reply: true,
      reply_text: "Bonjour",
      reason: "ok",
    });
    expect(result.allowReply).toBe(true);
    expect(result.aiStatus).toBe("auto_replied");
  });

  it("blocks recovery when confidence below threshold", () => {
    const result = applyReplyGate(NOT_INTERESTED_STATUS, {
      should_reply: true,
      reply_text: "Bonjour",
      reason: "recoverable",
      recovery_confidence: RECOVERY_CONFIDENCE_THRESHOLD - 1,
    });
    expect(result.allowReply).toBe(false);
    expect(result.aiStatus).toBe("skipped_recovery");
  });

  it("allows recovery at threshold", () => {
    const result = applyReplyGate(NOT_INTERESTED_STATUS, {
      should_reply: true,
      reply_text: "Bonjour",
      reason: "recoverable",
      recovery_confidence: RECOVERY_CONFIDENCE_THRESHOLD,
    });
    expect(result.allowReply).toBe(true);
  });

  it("abstains when should_reply is false", () => {
    const result = applyReplyGate(null, {
      should_reply: false,
      reply_text: null,
      reason: "refus définitif",
      recovery_confidence: 10,
    });
    expect(result.allowReply).toBe(false);
    expect(result.aiStatus).toBe("skipped_unsafe");
  });
});
